homeassistant/helpers/entity_component.py

```diff
@@ -189,41 +271,64 @@ def __init__(self, component, scan_interval, entity_namespace):
         self.scan_interval = scan_interval
         self.entity_namespace = entity_namespace
         self.platform_entities = []
-        self._unsub_polling = None
+        self._async_unsub_polling = None
 
     def add_entities(self, new_entities):
         """Add entities for a single platform."""
-        with self.component.lock:
-            for entity in new_entities:
-                if self.component.add_entity(entity, self):
-                    self.platform_entities.append(entity)
+        run_coroutine_threadsafe(
+            self.async_add_entities(new_entities), self.component.hass.loop
+        ).result()
 
-            self.component.update_group()
+    @asyncio.coroutine
+    def async_add_entities(self, new_entities):
+        """Add entities for a single platform async.
 
-            if self._unsub_polling is not None or \
-               not any(entity.should_poll for entity
-                       in self.platform_entities):
-                return
+        This method must be run in the event loop.
+        """
+        for entity in new_entities:
+            ret = yield from self.component.async_add_entity(entity, self)
+            if ret:
+                self.platform_entities.append(entity)
 
-            self._unsub_polling = track_utc_time_change(
-                self.component.hass, self._update_entity_states,
-                second=range(0, 60, self.scan_interval))
+        self.component.async_update_group()
 
-    def reset(self):
-        """Remove all entities and reset data."""
+        if self._async_unsub_polling is not None or \
+           not any(entity.should_poll for entity
+                   in self.platform_entities):
+            return
+
+        self._async_unsub_polling = async_track_utc_time_change(
+            self.component.hass, self._update_entity_states,
+            second=range(0, 60, self.scan_interval))
+
+    def async_reset(self):
+        """Remove all entities and reset data.
+
+        This method must be run in the event loop.
+        """
         for entity in self.platform_entities:
-            entity.remove()
-        if self._unsub_polling is not None:
-            self._unsub_polling()
-            self._unsub_polling = None
+            entity.async_remove()
```

Comment:
We don't know what the device will have to do when we remove it, thus this should be a coroutine